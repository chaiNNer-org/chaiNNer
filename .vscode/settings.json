{
  "files.insertFinalNewline": true,
  "files.trimFinalNewlines": false,
  "files.trimTrailingWhitespace": true,
  "python.analysis.typeCheckingMode": "basic",
  "cSpell.words": [
    "absdiff",
    "arange",
    "arctan",
    "argmax",
    "ascontiguousarray",
    "astype",
    "batchwise",
    "BSRGAN",
    "Chainner",
    "Chakra",
    "checkpointing",
    "codebook",
    "codeformer",
    "convolutional",
    "cvt",
    "denoising",
    "Denormalize",
    "depthwise",
    "downsample",
    "downsampling",
    "downscaled",
    "dstack",
    "dtype",
    "DXGI",
    "einops",
    "ESRGAN",
    "facexlib",
    "ffprobe",
    "fromarray",
    "frombuffer",
    "geforce",
    "GFPGAN",
    "hconcat",
    "hsl",
    "iinfo",
    "imdecode",
    "imencode",
    "imread",
    "imtype",
    "imwrite",
    "jit",
    "Lanczos",
    "layouted",
    "leakyrelu",
    "localstorage",
    "Luma",
    "maximizable",
    "meshgrid",
    "minimizable",
    "MSE",
    "navi",
    "ndim",
    "numel",
    "numpy",
    "onnxoptimizer",
    "onnxruntime",
    "OOM",
    "opencv",
    "pixelshuffle",
    "pixelshuffledirect",
    "pointwise",
    "prelu",
    "pretrained",
    "Prewitt",
    "protobuf",
    "PSNR",
    "quadro",
    "Quantizer",
    "randn",
    "reactflow",
    "relu",
    "restoreformer",
    "RRDB",
    "sanic",
    "Scharr",
    "setflags",
    "sobel",
    "SPSR",
    "SRGAN",
    "SSIM",
    "Swin",
    "tensorrt",
    "texconv",
    "timm",
    "tobytes",
    "torchscript",
    "torchvision",
    "Unembedding",
    "Unpickler",
    "Unsharp",
    "unsqueeze",
    "Upsample",
    "upsampler",
    "upsampling",
    "vconcat",
    "vkallocator",
    "vram",
    "vulkan",
    "YUV"
  ],
  "[python]": {
    "editor.defaultFormatter": "charliermarsh.ruff",
    "editor.formatOnSave": true,
    "editor.codeActionsOnSave": {
      "source.organizeImports": "explicit",
      "source.fixAll": "always"
    }
  },
  "stylelint.validate": [
    "css",
    "scss"
  ],
  "[scss]": {
    "editor.defaultFormatter": "stylelint.vscode-stylelint",
    "editor.formatOnSave": true,
  },
  "[typescript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.formatOnSave": true,
  },
  "[javascript]": {
    "editor.formatOnSave": true
  }
}
