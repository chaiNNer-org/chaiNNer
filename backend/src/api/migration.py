"""
Declarative per-node migrations for chaiNNer.

This module provides the infrastructure for declaring migrations on nodes themselves,
rather than as global migration functions in the frontend. This allows:
1. Migrations to be defined in Python alongside the nodes they affect
2. Trivial reuse of common migration patterns
3. Plugin authors to add migrations without frontend knowledge
4. Decoupling migrations from save data format

Migration types include:
- rename: Node was renamed (old schema_id -> new schema_id)
- change_inputs: Inputs were added, removed, reordered, or had their defaults changed
- change_outputs: Outputs were added, removed, or reordered
"""
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class Migration(ABC):
    """
    Base class for all migrations.
    
    Each migration represents a change to a node that requires updating save data.
    Migrations are applied in order, and may depend on specific versions of other nodes.
    """
    
    @abstractmethod
    def to_dict(self) -> dict[str, Any]:
        """Convert the migration to a JSON-serializable dictionary."""
        pass


@dataclass(frozen=True)
class RenameMigration(Migration):
    """
    Migration for when a node's schema_id has changed.
    
    Args:
        old: The old schema_id that should be migrated to the new one
    
    Example:
        @register(
            schema_id="chainner:image:my_new_name",
            migrations=[rename(old="chainner:image:my_old_name")]
        )
    """
    old: str
    
    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": "rename",
            "old": self.old,
        }


@dataclass(frozen=True)
class ChangeInputsMigration(Migration):
    """
    Migration for when a node's inputs have changed.
    
    This can handle:
    - Adding new inputs (they will use their default values)
    - Removing inputs (the input data will be discarded)
    - Renaming inputs (mapping old input IDs to new ones)
    - Changing input defaults
    
    Args:
        remove: List of input IDs to remove
        rename: Mapping of old input IDs to new input IDs
        add: Mapping of new input IDs to their default values
    
    Example:
        @register(
            migrations=[
                change_inputs(
                    rename={1: 2, 2: 1},  # swap inputs 1 and 2
                    remove=[3],  # remove old input 3
                )
            ]
        )
    """
    remove: list[int] | None = None
    rename: dict[int, int] | None = None
    add: dict[int, Any] | None = None
    
    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {"kind": "change_inputs"}
        if self.remove:
            result["remove"] = self.remove
        if self.rename:
            result["rename"] = {str(k): v for k, v in self.rename.items()}
        if self.add:
            result["add"] = {str(k): v for k, v in self.add.items()}
        return result


@dataclass(frozen=True)
class ChangeOutputsMigration(Migration):
    """
    Migration for when a node's outputs have changed.
    
    This can handle:
    - Removing outputs (edges will be removed)
    - Renaming outputs (mapping old output IDs to new ones)
    
    Args:
        remove: List of output IDs to remove
        rename: Mapping of old output IDs to new output IDs
    
    Example:
        @register(
            migrations=[
                change_outputs(
                    rename={0: 1, 1: 0},  # swap outputs
                )
            ]
        )
    """
    remove: list[int] | None = None
    rename: dict[int, int] | None = None
    
    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {"kind": "change_outputs"}
        if self.remove:
            result["remove"] = self.remove
        if self.rename:
            result["rename"] = {str(k): v for k, v in self.rename.items()}
        return result


@dataclass(frozen=True)
class NodeDependencyMigration(Migration):
    """
    Migration that depends on a specific version of another node.
    
    This is used when a migration needs to create or interact with another node,
    and that node's schema may have changed. By declaring a dependency on a specific
    version, the migration system ensures that:
    1. The dependency node is migrated to the specified version before this migration runs
    2. This migration is applied before the dependency node is migrated beyond the specified version
    
    Args:
        schema_id: The schema_id of the node this migration depends on
        version: The version of that node this migration was written for
    
    Example:
        If migration m2 of node M creates a node N v1, and N has 3 migrations (n1, n2, n3):
        
        @register(
            schema_id="M",
            migrations=[
                m1,
                depends_on("N", version=1),  # m2 will create N v1
                m2,  # creates N v1 nodes
                m3,
            ]
        )
        
        This ensures m2 runs after n1 but before n2, so created nodes will be N v1.
    """
    schema_id: str
    version: int
    
    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": "node_dependency",
            "schemaId": self.schema_id,
            "version": self.version,
        }


# Convenience functions for creating migrations
def rename(old: str) -> RenameMigration:
    """Create a rename migration."""
    return RenameMigration(old=old)


def change_inputs(
    remove: list[int] | None = None,
    rename: dict[int, int] | None = None,
    add: dict[int, Any] | None = None,
) -> ChangeInputsMigration:
    """Create a change_inputs migration."""
    return ChangeInputsMigration(remove=remove, rename=rename, add=add)


def change_outputs(
    remove: list[int] | None = None,
    rename: dict[int, int] | None = None,
) -> ChangeOutputsMigration:
    """Create a change_outputs migration."""
    return ChangeOutputsMigration(remove=remove, rename=rename)


def depends_on(schema_id: str, version: int) -> NodeDependencyMigration:
    """Create a node dependency migration."""
    return NodeDependencyMigration(schema_id=schema_id, version=version)
