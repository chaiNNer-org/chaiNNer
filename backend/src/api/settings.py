from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, TypedDict, Union

from sanic.log import logger

SettingsJson = Dict[str, object]
JsonExecutionOptions = Dict[str, SettingsJson]


class ExecutionOptions:
    def __init__(
        self,
        backend_settings: JsonExecutionOptions,
    ) -> None:
        self.__settings = backend_settings
        self.__parsers: dict[str, SettingsParser] = {}

        logger.info(f"Execution options: {self.__settings}")

    @staticmethod
    def parse(json: JsonExecutionOptions) -> ExecutionOptions:
        return ExecutionOptions(backend_settings=json)

    def get_package_settings_json(self, package_id: str) -> SettingsJson:
        return self.__settings.get(package_id, {})

    def get_package_settings(self, package_id: str) -> SettingsParser:
        parser = self.__parsers.get(package_id)
        if parser is None:
            parser = SettingsParser(self.get_package_settings_json(package_id))
            self.__parsers[package_id] = parser
        return parser


class SettingsParser:
    def __init__(self, raw: SettingsJson) -> None:
        self.__settings = raw

    def get_bool(self, key: str, default: bool) -> bool:
        value = self.__settings.get(key, default)
        if isinstance(value, bool):
            return value
        raise ValueError(f"Invalid bool value for {key}: {value}")

    def get_int(self, key: str, default: int, parse_str: bool = False) -> int:
        value = self.__settings.get(key, default)
        if parse_str and isinstance(value, str):
            return int(value)
        if isinstance(value, int) and not isinstance(value, bool):
            return value
        raise ValueError(f"Invalid str value for {key}: {value}")

    def get_str(self, key: str, default: str) -> str:
        value = self.__settings.get(key, default)
        if isinstance(value, str):
            return value
        raise ValueError(f"Invalid str value for {key}: {value}")

    def get_cache_location(self, key: str) -> str | None:
        value = self.__settings.get(key)
        if isinstance(value, str) or value is None:
            return value or None
        raise ValueError(f"Invalid cache location value for {key}: {value}")


@dataclass
class ToggleSetting:
    label: str
    key: str
    description: str
    default: bool = False
    disabled: bool = False
    type: str = "toggle"


class DropdownOption(TypedDict):
    label: str
    value: str


@dataclass
class DropdownSetting:
    label: str
    key: str
    description: str
    options: list[DropdownOption]
    default: str
    disabled: bool = False
    type: str = "dropdown"


@dataclass
class NumberSetting:
    label: str
    key: str
    description: str
    min: float
    max: float
    default: float = 0
    disabled: bool = False
    type: str = "number"


@dataclass
class CacheSetting:
    label: str
    key: str
    description: str
    directory: str
    default: str = ""
    disabled: bool = False
    type: str = "cache"


Setting = Union[ToggleSetting, DropdownSetting, NumberSetting, CacheSetting]
