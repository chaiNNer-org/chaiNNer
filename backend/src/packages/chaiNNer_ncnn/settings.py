from dataclasses import dataclass

try:
    from ncnn_vulkan import ncnn

    use_gpu = True
except ImportError:
    from ncnn import ncnn  # type: ignore

    use_gpu = False

from api import DropdownSetting, NodeContext, NumberSetting, ToggleSetting
from system import is_arm_mac

from . import package

if not is_arm_mac and use_gpu:
    try:
        gpu_list = []
        for i in range(ncnn.get_gpu_count()):
            gpu_list.append(ncnn.get_gpu_info(i).device_name())

        package.add_setting(
            DropdownSetting(
                label="GPU",
                key="gpu_index",
                description="Which GPU to use for NCNN. This is only relevant if you have multiple GPUs.",
                options=[{"label": x, "value": str(i)} for i, x in enumerate(gpu_list)],
                default="0",
            )
        )
    except Exception:
        pass

default_net_opt = ncnn.Net().opt

# Haven't tested disabling Winograd/SGEMM in the ncnn_vulkan fork, so only
# allow it with upstream ncnn for now. It should work fine regardless of
# CPU/GPU, but I only tested with CPU. Ditto for multithreading, except it
# only makes sense for CPU.
if not use_gpu:
    package.add_setting(
        ToggleSetting(
            label="Use Winograd",
            key="winograd",
            description="Enable Winograd convolution for NCNN. Typically faster but uses more memory.",
            default=default_net_opt.use_winograd_convolution,
        )
    )
    package.add_setting(
        ToggleSetting(
            label="Use SGEMM",
            key="sgemm",
            description="Enable SGEMM convolution for NCNN. Typically faster but uses more memory.",
            default=default_net_opt.use_sgemm_convolution,
        )
    )

    package.add_setting(
        NumberSetting(
            label="Thread Count",
            key="threads",
            description="Number of threads for NCNN. Only affects CPU mode.",
            default=default_net_opt.num_threads,
            min=1,
            max=default_net_opt.num_threads,
        )
    )
    package.add_setting(
        NumberSetting(
            label="Block Time",
            key="blocktime",
            description="Milliseconds for threads to busy-wait for more work before going to sleep. Higher values may be faster; lower values may decrease power consumption and CPU usage. Only affects CPU mode.",
            default=default_net_opt.openmp_blocktime,
            min=0,
            max=400,
        )
    )

package.add_setting(
    NumberSetting(
        label="Memory Budget Limit (GiB)",
        key="budget_limit",
        description="Maximum memory to use for NCNN inference. 0 means no limit. Memory usage measurement is not completely accurate yet; you may need to significantly adjust this budget limit via trial-and-error if it's not having the effect you want.",
        default=0,
        min=0,
        max=1024**2,
    )
)


@dataclass(frozen=True)
class NcnnSettings:
    gpu_index: int
    winograd: bool
    sgemm: bool
    threads: int
    blocktime: int
    budget_limit: int


def get_settings(context: NodeContext) -> NcnnSettings:
    settings = context.settings

    return NcnnSettings(
        gpu_index=settings.get_int("gpu_index", 0, parse_str=True),
        winograd=settings.get_bool(
            "winograd", default_net_opt.use_winograd_convolution
        ),
        sgemm=settings.get_bool("sgemm", default_net_opt.use_sgemm_convolution),
        threads=settings.get_int(
            "threads", default_net_opt.num_threads, parse_str=True
        ),
        blocktime=settings.get_int(
            "blocktime", default_net_opt.openmp_blocktime, parse_str=True
        ),
        budget_limit=settings.get_int("budget_limit", 0, parse_str=True),
    )
