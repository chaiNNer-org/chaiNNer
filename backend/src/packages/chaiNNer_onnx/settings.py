from __future__ import annotations

import os
from dataclasses import dataclass
from typing import List, cast

import onnxruntime as ort
from sanic.log import logger

from api import CacheSetting, DropdownSetting, NodeContext, ToggleSetting
from gpu import get_nvidia_helper
from system import is_arm_mac

from . import package

nv = get_nvidia_helper()

if not is_arm_mac:
    gpu_list = nv.list_gpus() if nv is not None else []

    package.add_setting(
        DropdownSetting(
            label="GPU",
            key="gpu_index",
            description="Which GPU to use for ONNX. This is only relevant if you have multiple GPUs.",
            options=[{"label": x, "value": str(i)} for i, x in enumerate(gpu_list)],
            default="0",
        )
    )


def get_providers():
    providers = cast(List[str], ort.get_available_providers())

    default = providers[0]
    if "CUDAExecutionProvider" in providers:
        default = "CUDAExecutionProvider"
    elif "CPUExecutionProvider" in providers:
        default = "CPUExecutionProvider"

    return providers, default


execution_providers, default_provider = get_providers()

if not is_arm_mac:
    package.add_setting(
        DropdownSetting(
            label="Execution Provider",
            key="execution_provider",
            description="What provider to use for ONNX.",
            options=[
                {"label": x.replace("ExecutionProvider", ""), "value": x}
                for x in execution_providers
            ],
            default=default_provider,
        )
    )

    package.add_setting(
        CacheSetting(
            label="Cache TensorRT Engines",
            key="onnx_tensorrt_cache",
            description="Whether to cache the converted TensorRT engines. This can significantly speed up subsequent runs.",
            directory="onnx_tensorrt_cache",
            disabled="TensorrtExecutionProvider" not in execution_providers,
        )
    )

    should_fp16 = False
    if nv is not None:
        should_fp16 = nv.supports_fp16()

    package.add_setting(
        ToggleSetting(
            label="Use TensorRT FP16 Mode",
            key="tensorrt_fp16_mode",
            description="Runs TensorRT in half-precision (FP16) mode for less VRAM usage. RTX GPUs also get a speedup.",
            default=should_fp16,
            disabled="TensorrtExecutionProvider" not in execution_providers,
        )
    )


@dataclass(frozen=True)
class OnnxSettings:
    gpu_index: int
    execution_provider: str
    tensorrt_cache_path: str | None
    tensorrt_fp16_mode: bool


def get_settings(context: NodeContext) -> OnnxSettings:
    settings = context.settings

    tensorrt_cache_path = settings.get_cache_location("onnx_tensorrt_cache")
    logger.info(f"TensorRT cache location: {tensorrt_cache_path}")
    if tensorrt_cache_path and not os.path.exists(tensorrt_cache_path):
        os.makedirs(tensorrt_cache_path)

    return OnnxSettings(
        gpu_index=settings.get_int("gpu_index", 0, parse_str=True),
        execution_provider=settings.get_str("execution_provider", default_provider),
        tensorrt_cache_path=tensorrt_cache_path,
        tensorrt_fp16_mode=settings.get_bool("tensorrt_fp16_mode", False),
    )
