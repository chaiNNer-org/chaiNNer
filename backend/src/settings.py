from dataclasses import dataclass
from typing import Any

from api import SettingsParser, get_execution_options


@dataclass(frozen=True)
class GeneralSettings:
    example: bool


def get_global_settings() -> Any:
    settings = SettingsParser(get_execution_options().get_package_settings("general"))

    return GeneralSettings(
        example=settings.get_bool("example", default=False),
    )
