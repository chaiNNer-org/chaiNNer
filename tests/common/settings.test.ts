import { expect, test } from 'vitest';
import { migrateOldStorageSettings } from '../../src/common/settings/migration';

const oldSettingData: Partial<Record<string, string>> = {
    'allow-multiple-instances': 'false',
    'animate-chain': 'true',
    'backend-settings':
        '{"chaiNNer_pytorch":{"gpu_index":"0","use_cpu":false,"use_fp16":false,"budget_limit":0},"chaiNNer_ncnn":{"gpu_index":"0","budget_limit":0},"chaiNNer_onnx":{"gpu_index":"0","execution_provider":"CUDAExecutionProvider","onnx_tensorrt_cache":"","tensorrt_fp16_mode":false}}',
    'check-upd-on-strtup': 'true',
    'check-upd-on-strtup-2': 'true',
    'disable-hw-accel': 'true',
    'enable-hardware-acceleration': 'false',
    'experimental-features': 'true',
    'is-cpu': 'false',
    'is-fp16': 'false',
    'ncnn-gpu': '0',
    'node-favorites':
        '["chainner:image:preview","chainner:image:save","chainner:image:average_color_fix","chainner:image:overlay","chainner:image:fill_alpha","chainner:pytorch:load_model","chainner:pytorch:upscale_image","chainner:utility:text_pattern","chainner:image:load","chainner:image:view","chainner:utility:copy_to_clipboard"]',
    'node-hidden': '[]',
    'node-selector-collapsed': 'true',
    'node-visibility-mode-active': 'true',
    'onnx-execution-provider': '"CUDAExecutionProvider"',
    'onnx-gpu': '0',
    'onnx-should-tensorrt-cache': 'false',
    'onnx-should-tensorrt-fp16': 'false',
    'pytorch-gpu': '0',
    'snap-to-grid': 'true',
    'snap-to-grid-amount': '16',
    'startup-template': '""',
    'system-python-location': '"C:\\\\python39\\\\python.exe"',
    theme: '"dark"',
    'use-dark-mode': 'true',
    'use-discord-rpc': 'true',
    'use-last-directory-chainner:image:file_iterator 0': 'C:\\DS3TexUp\\upscale',
    'use-last-directory-chainner:image:file_iterator-0':
        'C:\\Users\\micha\\Desktop\\detail n\\Neuer Ordner',
    'use-last-directory-chainner:image:load 0': 'C:\\Users\\micha\\Desktop',
    'use-last-directory-chainner:image:load-0': 'C:\\Users\\micha\\Desktop',
    'use-last-directory-chainner:image:load_images-0': 'C:\\Users\\micha\\Desktop\\Irithyll',
    'use-last-directory-chainner:image:load_image_pairs-0': 'C:\\DS3TexUp\\extract\\chr',
    'use-last-directory-chainner:image:load_image_pairs-1': 'C:\\DS3TexUp',
    'use-last-directory-chainner:image:load_video-0': 'D:\\Videos\\games',
    'use-last-directory-chainner:image:save 1': 'C:\\DS3TexUp\\up-manual-new',
    'use-last-directory-chainner:image:save-1': 'C:\\Users\\micha\\Desktop',
    'use-last-directory-chainner:image:save_video-1': 'C:\\Users\\micha\\Desktop',
    'use-last-directory-chainner:ncnn:load_model 0':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\.ncnn-models\\4x_RealSR_DF2K_JPEG',
    'use-last-directory-chainner:ncnn:load_model 1':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\.ncnn-models\\4x_RealSR_DF2K_JPEG',
    'use-last-directory-chainner:ncnn:load_model-0':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN',
    'use-last-directory-chainner:ncnn:load_model-1':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN\\realesrgan-x4plus.ncnn',
    'use-last-directory-chainner:ncnn:save_model-1':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\.ncnn-models\\4x_RealSR_DF2K_JPEG',
    'use-last-directory-chainner:onnx:load_model 0':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN\\onnx',
    'use-last-directory-chainner:onnx:load_model-0': 'C:\\Users\\micha\\Downloads',
    'use-last-directory-chainner:onnx:save_model 1':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN\\onnx',
    'use-last-directory-chainner:onnx:save_model-1': 'C:\\Users\\micha\\Desktop',
    'use-last-directory-chainner:pytorch:load_data-0':
        'C:\\Users\\micha\\Git\\model-database\\data\\models',
    'use-last-directory-chainner:pytorch:load_model 0':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN',
    'use-last-directory-chainner:pytorch:load_model-0': 'C:\\Users\\micha\\Downloads',
    'use-last-directory-chainner:pytorch:load_models-0': 'C:\\DS3TexUp\\extract\\m31',
    'use-last-directory-chainner:pytorch:model_file_iterator-0':
        'C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN',
    'use-last-directory-chainner:pytorch:save_model 1': 'C:\\Users\\micha\\Desktop',
    'use-last-directory-chainner:utility:directory-0': 'C:\\DS3TexUp\\extract\\chr',
    'use-last-window-size': '{"maximized":false,"width":1527,"height":1007}',
    'use-recently-open':
        '["C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\tile.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\sliders test.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\test.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\simple-upscale.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\normal-pattern-add.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\normal.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\key color upscale.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\asdasdsadasd.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\_resize test.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\sfx-workflow.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\gaussian research.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\ds-normal-test.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\wood.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\temp.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\split combined lr sr.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\_save vid test.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\_bp.chn","C:\\\\Users\\\\micha\\\\Downloads\\\\Batch_Upscale_Icons.chn","C:\\\\Users\\\\micha\\\\Desktop\\\\chains\\\\crash chain.chn"]',
    'use-system-python': 'false',
    'viewport-export-padding': '20',
};

test(`Migrate settings`, () => {
    const unusedKeys = new Set(Object.keys(oldSettingData));
    const settings = migrateOldStorageSettings({
        keys: Object.keys(oldSettingData),
        getItem: (key: string) => {
            unusedKeys.delete(key);
            return oldSettingData[key] ?? null;
        },
    });

    expect(settings).toMatchSnapshot();
    expect(unusedKeys).toMatchSnapshot();
});
