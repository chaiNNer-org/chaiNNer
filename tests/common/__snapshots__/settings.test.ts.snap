// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Migrate settings 1`] = `
{
  "allowMultipleInstances": false,
  "animateChain": true,
  "checkForUpdatesOnStartup": true,
  "experimentalFeatures": true,
  "favoriteNodes": [
    "chainner:image:preview",
    "chainner:image:save",
    "chainner:image:average_color_fix",
    "chainner:image:overlay",
    "chainner:image:fill_alpha",
    "chainner:pytorch:load_model",
    "chainner:pytorch:upscale_image",
    "chainner:utility:text_pattern",
    "chainner:image:load",
    "chainner:image:view",
    "chainner:utility:copy_to_clipboard",
  ],
  "hardwareAcceleration": false,
  "lastWindowSize": {
    "height": 1007,
    "maximized": false,
    "width": 1527,
  },
  "packageSettings": {
    "chaiNNer_ncnn": {
      "budget_limit": 0,
      "gpu_index": "0",
    },
    "chaiNNer_onnx": {
      "execution_provider": "CUDAExecutionProvider",
      "gpu_index": "0",
      "onnx_tensorrt_cache": "",
      "tensorrt_fp16_mode": false,
    },
    "chaiNNer_pytorch": {
      "budget_limit": 0,
      "gpu_index": "0",
      "use_cpu": false,
      "use_fp16": false,
    },
  },
  "snapToGrid": true,
  "snapToGridAmount": 16,
  "startupTemplate": "",
  "storage": {
    "lastDirectories": {
      "chainner:image:file_iterator-0": "C:\\Users\\micha\\Desktop\\detail n\\Neuer Ordner",
      "chainner:image:load-0": "C:\\Users\\micha\\Desktop",
      "chainner:image:load_image_pairs-0": "C:\\DS3TexUp\\extract\\chr",
      "chainner:image:load_image_pairs-1": "C:\\DS3TexUp",
      "chainner:image:load_images-0": "C:\\Users\\micha\\Desktop\\Irithyll",
      "chainner:image:load_video-0": "D:\\Videos\\games",
      "chainner:image:save-1": "C:\\Users\\micha\\Desktop",
      "chainner:image:save_video-1": "C:\\Users\\micha\\Desktop",
      "chainner:ncnn:load_model-0": "C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN",
      "chainner:ncnn:load_model-1": "C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN\\realesrgan-x4plus.ncnn",
      "chainner:ncnn:save_model-1": "C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\.ncnn-models\\4x_RealSR_DF2K_JPEG",
      "chainner:onnx:load_model-0": "C:\\Users\\micha\\Downloads",
      "chainner:onnx:save_model-1": "C:\\Users\\micha\\Desktop",
      "chainner:pytorch:load_data-0": "C:\\Users\\micha\\Git\\model-database\\data\\models",
      "chainner:pytorch:load_model-0": "C:\\Users\\micha\\Downloads",
      "chainner:pytorch:load_models-0": "C:\\DS3TexUp\\extract\\m31",
      "chainner:pytorch:model_file_iterator-0": "C:\\DS3TexUp\\Cupscale 1.39.0f1\\CupscaleData\\models\\ESRGAN",
      "chainner:utility:directory-0": "C:\\DS3TexUp\\extract\\chr",
    },
    "nodeSelectorCollapsed": true,
    "recent": [
      "C:\\Users\\micha\\Desktop\\chains\\tile.chn",
      "C:\\Users\\micha\\Desktop\\chains\\sliders test.chn",
      "C:\\Users\\micha\\Desktop\\chains\\test.chn",
      "C:\\Users\\micha\\Desktop\\chains\\simple-upscale.chn",
      "C:\\Users\\micha\\Desktop\\chains\\normal-pattern-add.chn",
      "C:\\Users\\micha\\Desktop\\chains\\normal.chn",
      "C:\\Users\\micha\\Desktop\\chains\\key color upscale.chn",
      "C:\\Users\\micha\\Desktop\\chains\\asdasdsadasd.chn",
      "C:\\Users\\micha\\Desktop\\chains\\_resize test.chn",
      "C:\\Users\\micha\\Desktop\\chains\\sfx-workflow.chn",
      "C:\\Users\\micha\\Desktop\\chains\\gaussian research.chn",
      "C:\\Users\\micha\\Desktop\\chains\\ds-normal-test.chn",
      "C:\\Users\\micha\\Desktop\\chains\\wood.chn",
      "C:\\Users\\micha\\Desktop\\temp.chn",
      "C:\\Users\\micha\\Desktop\\chains\\split combined lr sr.chn",
      "C:\\Users\\micha\\Desktop\\chains\\_save vid test.chn",
      "C:\\Users\\micha\\Desktop\\chains\\_bp.chn",
      "C:\\Users\\micha\\Downloads\\Batch_Upscale_Icons.chn",
      "C:\\Users\\micha\\Desktop\\chains\\crash chain.chn",
    ],
  },
  "systemPythonLocation": "",
  "theme": "dark",
  "useSystemPython": false,
  "viewportExportPadding": 20,
}
`;

exports[`Migrate settings 2`] = `
Set {
  "check-upd-on-strtup",
  "disable-hw-accel",
  "is-cpu",
  "is-fp16",
  "ncnn-gpu",
  "node-hidden",
  "node-visibility-mode-active",
  "onnx-execution-provider",
  "onnx-gpu",
  "onnx-should-tensorrt-cache",
  "onnx-should-tensorrt-fp16",
  "pytorch-gpu",
  "system-python-location",
  "use-dark-mode",
  "use-discord-rpc",
  "use-last-directory-chainner:image:file_iterator 0",
  "use-last-directory-chainner:image:load 0",
  "use-last-directory-chainner:image:save 1",
  "use-last-directory-chainner:ncnn:load_model 0",
  "use-last-directory-chainner:ncnn:load_model 1",
  "use-last-directory-chainner:onnx:load_model 0",
  "use-last-directory-chainner:onnx:save_model 1",
  "use-last-directory-chainner:pytorch:load_model 0",
  "use-last-directory-chainner:pytorch:save_model 1",
}
`;
